File: services/circuit-agent/src/infra/search/OpenRouterSearch.ts
import type { SearchProvider } from '../../domain/contracts/index.js'
import { OpenRouterTextProvider } from '../providers/OpenRouterTextProvider.js'
import { logger } from '../log/logger.js'

// 中文注释：基于 OpenRouter 的在线检索 provider，使用 openrouter 的 :online 模型
export class OpenRouterSearch implements SearchProvider {
  private provider: OpenRouterTextProvider
  private model: string
  private headers?: Record<string,string>
  private trace?: (evt: any) => Promise<void> | void

  constructor(baseUrl: string, defaultTimeoutMs: number, headers?: Record<string,string>, options?: { modelOverride?: string, forceOnline?: boolean, trace?: (evt: any) => Promise<void> | void }) {
    this.provider = new OpenRouterTextProvider(baseUrl, defaultTimeoutMs)
    const configured = String(options?.modelOverride || process.env.OPENROUTER_SEARCH_MODEL || '').trim()
    const fallback = 'perplexity/sonar'
    const raw = configured || fallback
    const forceOnline = (options?.forceOnline === true) || (String(process.env.OPENROUTER_FORCE_ONLINE || '').trim() === '1')
    // 若是 Perplexity 系列，一般自带联网能力，无需 :online；其它模型在需要时追加 :online 后缀
    const needsOnlineSuffix = forceOnline && !/perplexity\//i.test(raw) && !/:online$/i.test(raw)
    this.model = needsOnlineSuffix ? `${raw}:online` : raw
    this.headers = headers
    this.trace = options?.trace
  }

  // query -> 调用 openrouter 在线模型，要求返回 JSON 数组 [{title,url}] 或以行为单位的条目
  async search(query: string, topN: number): Promise<{ title: string; url: string }[]> {
    try {
      logger.info('web.search.start', { query: query.slice(0, 120), topN })
      const system = `You are a web search tool. Given the user query, return a JSON array of up to ${topN} items, each an object with keys \"title\" and \"url\". Prefer authoritative sources (original vendors, datasheets, application notes). Return only the JSON array and no extra text.`
      const userMsg = query
      // 显式启用 web 插件进行真实搜索
      const engine = String(process.env.OPENROUTER_WEB_ENGINE || 'exa')
      const plugins = [{ id: 'web', engine, max_results: Math.max(1, Number(topN || 5)) }]
      // 记录原始请求
      try { await this.trace?.({ phase: 'search', target: 'query', direction: 'request', meta: { model: this.model, engine, topN }, body: { system, messages: [{ role: 'user', content: userMsg }], plugins } }) } catch {}
      const resp = await this.provider.chat({ apiUrl: '', model: this.model, system, messages: [{ role: 'user', content: userMsg }], plugins, headers: this.headers })
      const txt = (resp && resp.text) ? resp.text.trim() : ''
      // 记录原始响应
      try { await this.trace?.({ phase: 'search', target: 'query', direction: 'response', meta: { model: this.model }, body: { raw: resp?.raw || '', text: txt } }) } catch {}
      logger.info('web.search.done', { length: txt.length })
      if (!txt) return []
      // 尝试解析为 JSON
      try {
        const j = JSON.parse(txt)
        if (Array.isArray(j)) {
          const out = j.slice(0, topN).map((it: any) => ({ title: String(it.title || it.name || '').trim(), url: String(it.url || it.link || '').trim() })).filter((x: any) => x.title && x.url)
          return out
        }
      } catch {}

      // 回退：尝试按行抽取 url/title（简单的 'title - url' 或 'url title' 格式）
      const lines = txt.split(/\r?\n/).map((l) => l.trim()).filter(Boolean)
      const results: { title: string; url: string }[] = []
      for (const l of lines) {
        if (results.length >= topN) break
        // 尝试 url 在行内
        const urlMatch = l.match(/(https?:\/\/[^\s,;]+)/i)
        if (urlMatch) {
          const url = urlMatch[1]
          const title = l.replace(url, '').replace(/[\-–—:]+/g, ' ').trim() || url
          results.push({ title, url })
        }
      }
      const out = results.slice(0, topN)
      logger.info('web.search.parsed', { count: out.length })
      return out
    } catch (e) {
      logger.warn('web.search.error', { message: (e as Error)?.message })
      return []
    }
  }

  // 中文注释：对 URL 进行在线抓取并在上游汇总为指定语言与词数上限的纯文本摘要
  async summarizeUrl(url: string, wordLimit: number, lang: 'zh'|'en'): Promise<string> {
    try {
      logger.info('web.summary.start', { url, wordLimit, lang })
      const limit = Math.max(64, Math.min(2048, Number(wordLimit || 512)))
      const system = `You are a web reader and summarizer. Fetch the given URL and write a concise summary in ${lang === 'zh' ? 'Chinese' : 'English'} within ${limit} words. Return plain text only.`
      const userMsg = `URL: ${url}`
      // 显式启用 web 插件；engine 可通过环境变量覆盖（native/exa）
      const engine = String(process.env.OPENROUTER_WEB_ENGINE || 'exa')
      const maxResults = Number(process.env.OPENROUTER_WEB_MAX_RESULTS || 1)
      const plugins = [{ id: 'web', engine, max_results: maxResults, search_prompt: 'Some relevant web results:' }]
      // 记录原始请求
      try { await this.trace?.({ phase: 'search', target: 'summary', direction: 'request', meta: { model: this.model, engine, url, wordLimit: limit, lang }, body: { system, messages: [{ role: 'user', content: userMsg }], plugins } }) } catch {}
      const resp = await this.provider.chat({ apiUrl: '', model: this.model, system, messages: [{ role: 'user', content: userMsg }], plugins, headers: this.headers })
      const txt = (resp && resp.text) ? resp.text.trim() : ''
      // 记录原始响应
      try { await this.trace?.({ phase: 'search', target: 'summary', direction: 'response', meta: { model: this.model, url }, body: { raw: resp?.raw || '', text: txt } }) } catch {}
      logger.info('web.summary.done', { length: txt.length })
      return txt
    } catch (e) {
      logger.warn('web.summary.error', { url, message: (e as Error)?.message })
      return ''
    }
  }
}

File: services/circuit-agent-py/app/infra/search/openrouter_search.py
from typing import List, Dict, Optional
from app.utils.prompt_loader import PromptLoader, PromptLoadError
from app.infra.http.openrouter_client import post_json
from app.infra.providers.text import OpenRouterTextProvider
from app.core.config import get_config

class OpenRouterSearch:
    def __init__(self, base_url: str, default_timeout_ms: int, headers: Optional[Dict[str,str]]=None, options: Optional[Dict]=None):
        self.provider = OpenRouterTextProvider(base_url, default_timeout_ms)
        configured = (options or {}).get('modelOverride') or ''
        fallback = 'perplexity/sonar'
        raw = configured or fallback
        force_online = (options or {}).get('forceOnline', False)
        needs_online = force_online and 'perplexity/' not in raw and not raw.endswith(':online')
        self.model = raw + (':online' if needs_online else '')
        self.headers = headers

    async def search(self, query: str, top_n: int) -> List[Dict[str,str]]:
        try:
            psearch = PromptLoader.load_prompt('circuit-agent', 'pass', 'zh', 'search')
            system = str(psearch).replace('{topN}', str(top_n))
        except PromptLoadError:
            system = ''
        user_msg = query
        plugins = [{ 'id': 'web', 'engine': 'exa', 'max_results': max(1, int(top_n or 5)) }]
        resp = await self.provider.chat('', self.model, system, [{ 'role': 'user', 'content': user_msg }], plugins=plugins, headers=self.headers)
        txt = (resp.get('text') or '').strip()
        if not txt:
            return []
        import json
        try:
            j = json.loads(txt)
            if isinstance(j, list):
                out = []
                for it in j[:top_n]:
                    title = str(it.get('title') or it.get('name') or '').strip()
                    url = str(it.get('url') or it.get('link') or '').strip()
                    if title and url:
                        out.append({'title': title, 'url': url})
                return out
        except Exception:
            pass
        lines = [l.strip() for l in txt.splitlines() if l.strip()]
        results = []
        for l in lines:
            if len(results) >= top_n:
                break
            import re
            m = re.search(r"(https?://[^\s,;]+)", l)
            if m:
                url = m.group(1)
                title = l.replace(url, '').replace('-', ' ').strip() or url
                results.append({'title': title, 'url': url})
        return results[:top_n]

    async def summarize_url(self, url: str, word_limit: int, lang: str) -> str:
        try:
            psummary = PromptLoader.load_prompt('circuit-agent', 'pass', 'zh', 'summary')
            system = str(psummary).replace('{limit}', str(max(64, min(2048, int(word_limit or 1024))))).replace('{lang}', 'Chinese' if lang == 'zh' else 'English')
        except PromptLoadError:
            system = ''
        user_msg = f"URL: {url}"
        model_name = 'qwen/qwen2.5-vl-72b-instruct:free'
        plugins = [{ 'id': 'web', 'engine': 'exa', 'max_results': int(1) }]
        resp = await self.provider.chat('', model_name, system, [{ 'role': 'user', 'content': user_msg }], plugins=plugins, headers=self.headers)
        txt = (resp.get('text') or '').strip()
        return txt
